ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+    def __init__(self, data, seq_length):
+        self.data = data
+        self.seq_length = seq_length
+
+    def __len__(self):
+        return len(self.data) - self.seq_length
+
+    def __getitem__(self, index):
+        X = self.data[index:index + self.seq_length]
+        y = self.data[index + self.seq_length]
+        return torch.tensor(X, dtype=torch.float32), torch.tensor(y, dtype=torch.float32)
+
+seq_length = 24
+train_dataset = TimeSeriesDataset(train_data_scaled, seq_length)
+test_dataset = TimeSeriesDataset(test_data_scaled, seq_length)
+
+# Create DataLoader
+train_loader = DataLoader(train_dataset, batch_size=8, shuffle=True)
+test_loader = DataLoader(test_dataset, batch_size=8, shuffle=False)
+```
+
+## LSTM Architecture
+
+We define our LSTM architecture in the code cell below.
+
+```{python}
+#| code-fold: false
+# Define the LSTM model
+class LSTMModel(nn.Module):
+    def __init__(self, input_size=1, hidden_size=16, output_size=1):
+        super(LSTMModel, self).__init__()
+        self.lstm = nn.LSTM(input_size, hidden_size, batch_first=True)
+        self.fc = nn.Linear(hidden_size, output_size)
+
+    def forward(self, x):
+        x, _ = self.lstm(x)
+        x = self.fc(x[:, -1, :])  # Use the output of the last time step
+        return x
+
+model = LSTMModel()
+print(model)
+```
+
+## Loss and Training
+
+We use a mean-square error loss function and the Adams optimizer. We train the model for 20 epochs. We display the decrease in the loss during training.
+
+```{python}
+# Set criterion and optimizer
+criterion = nn.MSELoss()
+optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+
+epochs = 20
+train_losses = []
+for epoch in range(epochs):
+    model.train()
+    train_loss = 0.0
+    for X, y in train_loader:
+
+      optimizer.zero_grad()
+      outputs = model(X)
+      loss = criterion(outputs, y)
+      loss.backward()
+      optimizer.step()
+
+      train_loss += loss.item()
+    train_losses.append(train_loss)
+
+# Plot the training losses over epochs
+plt.figure(figsize=(10, 5))
+plt.plot(range(1, epochs + 1), train_losses, marker='o', linestyle='-', color='b')
+plt.title('Training Loss over Epochs')
+plt.xlabel('Epochs')
+plt.ylabel('Loss')
+plt.xticks(range(1, epochs + 1))  # Set x-tick marks to be integers
+plt.grid(True)
+plt.show()
+```
+
+## Evaluation
+
+Let's evaluate the model's performance.
+
+```{python}
+# Evaluate the model
+model.eval()
+predictions = []
+trues = []
+with torch.no_grad():
+    for X, y in test_loader:
+        preds = model(X)
+        predictions.extend(preds.numpy())
+        trues.extend(y.numpy())
+
+# Rescale predictions and true to original scale
+predictions_rescaled = scaler.inverse_transform(predictions)
+true_rescaled = scaler.inverse_transform(trues)
+
+# Plot results
+plt.figure(figsize=(10, 6))
+plt.plot(true_rescaled, label='True Values')
+plt.plot(predictions_rescaled, label='Predicted Values', alpha=0.7)
+plt.legend()
+plt.show()
+```
+
```

Comment:
Can we add some interpretations of this plot? Another opportunity for class input. Did it do well? Any obvious problems?

Reply:
Added questions to prompt discussion.